server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java

```diff
@@ -141,144 +144,150 @@ static boolean isIdleThread(ThreadInfo threadInfo) {
-        sb.append(interval);
-        sb.append(", busiestThreads=");
-        sb.append(busiestThreads);
-        sb.append(", ignoreIdleThreads=");
-        sb.append(ignoreIdleThreads);
-        sb.append(":\n");
-
-        Map<Long, ThreadTimeAccumulator> threadInfos = new HashMap<>();
-        for (long threadId : threadBean.getAllThreadIds()) {
-            // ignore our own thread...
-            if (currentThreadId == threadId) {
+        for (int i = 0; i < threadIds.length; i++) {
+            if (threadInfos[i] == null || threadIds[i] == currentThreadId) {
                 continue;
             }
-            long cpu = threadBean.getThreadCpuTime(threadId);
-            if (cpu == -1) {
+            long cpuTime = threadBean.getThreadCpuTime(threadIds[i]);
+            if (cpuTime == INVALID_TIMING) {
                 continue;
             }
-            ThreadInfo info = threadBean.getThreadInfo(threadId, 0);
-            if (info == null) {
-                continue;
-            }
-            threadInfos.put(threadId, new ThreadTimeAccumulator(info, cpu));
+            result.put(threadIds[i], new ThreadTimeAccumulator(threadInfos[i], cpuTime));
         }
-        Thread.sleep(interval.millis());
-        for (long threadId : threadBean.getAllThreadIds()) {
-            // ignore our own thread...
-            if (currentThreadId == threadId) {
-                continue;
-            }
-            long cpu = threadBean.getThreadCpuTime(threadId);
-            if (cpu == -1) {
-                threadInfos.remove(threadId);
-                continue;
-            }
-            ThreadInfo info = threadBean.getThreadInfo(threadId, 0);
-            if (info == null) {
-                threadInfos.remove(threadId);
-                continue;
-            }
-            ThreadTimeAccumulator data = threadInfos.get(threadId);
-            if (data != null) {
-                data.setDelta(info, cpu);
-            } else {
-                threadInfos.remove(threadId);
-            }
-        }
-        // sort by delta CPU time on thread.
-        List<ThreadTimeAccumulator> hotties = new ArrayList<>(threadInfos.values());
-        final int busiestThreads = Math.min(this.busiestThreads, hotties.size());
-        // skip that for now
-        final ToLongFunction<ThreadTimeAccumulator> getter = ThreadTimeAccumulator.valueGetterForReportType(type);
-
-        CollectionUtil.introSort(hotties, Comparator.comparingLong(getter).reversed());
-
-        // analyse N stack traces for M busiest threads
-        long[] ids = new long[busiestThreads];
-        for (int i = 0; i < busiestThreads; i++) {
-            ThreadTimeAccumulator info = hotties.get(i);
-            ids[i] = info.info.getThreadId();
-        }
-        ThreadInfo[][] allInfos = new ThreadInfo[threadElementsSnapshotCount][];
+
+        return result;
+    }
+
+    ThreadInfo[][] captureThreadStacks(ThreadMXBean threadBean, long[] threadIds) throws InterruptedException {
```

Comment:
This method is a simple extraction from innerDetect, so that we can test it separately and reduce the code inside innerDetect. Reducing the the main-line code in innerDetect helps with clarity around the try - finally block for enabling/disabling thread contention monitoring.